### ISSUE TITLE
Optimizing useForm in Inertia.js: Preventing Unnecessary Re-renders

### ISSUE THREAD

**Issue by HassanDev13** [inertiajs/inertia#2084]
### Version:

- `@inertiajs/react` version: 1.0.0

### Describe the problem:

I'm encountering a performance challenge with the useForm hook in Inertia.js while rendering a form with initial data in the following structure:

```
{
  a: 0,
  b: 0,
  c: true
}
```
The issue arises when modifying one input component; it triggers a re-render of all form components. While this behavior is manageable with small forms, it poses a performance concern for larger, more complex forms.

Could you provide guidance on optimizing useForm to prevent unnecessary re-renders or suggest any best practices to improve performance in this context?

### Steps to reproduce:

1 - Set up Inertia.js with React:

Install Inertia.js and @inertiajs/react version 1.0.0 in your project.
Create a simple page component using useForm.

2 - Define Initial Form State:

Use useForm to initialize form data with a structure like the following:

```
const { data, setData } = useForm({
  a: 0,
  b: 0,
  c: true
});
```
3 - Render Multiple Input Components:

Set up three input components to bind each of the fields (a, b, c) from the form state, for example:

```
<input
  type="number"
  value={data.a}
  onChange={(e) => setData('a', e.target.value)}
/>
<input
  type="number"
  value={data.b}
  onChange={(e) => setData('b', e.target.value)}
/>
<input
  type="checkbox"
  checked={data.c}
  onChange={(e) => setData('c', e.target.checked)}
/>
```
4 - Observe Component Re-renders:

Use React DevTools to monitor re-renders. Change the value of one input (e.g., data.a) and observe that all input components are re-rendered each time an individual field updates.

5 - Test with Additional Fields:

To simulate a complex form, expand the initial form state with additional fields (e.g., d, e, f, etc.) and repeat step 4. The re-renders should increase, causing noticeable performance degradation as the form grows.

**Comment by Hasan-Mir** [comment]
This behavior is due to React, not Inertia. To prevent unnecessary re-renders, consider wrapping the inputs in a custom `Input` component. Memoize this component with `React.memo` and use `useCallback` for the `onChange` handlers.

**Comment by HassanDev13** [comment]
True, it’s a React issue, but the `useForm` hook provided by Inertia can handle this similarly to how [React Hook Form](https://react-hook-form.com/docs/useform) does.

**Comment by pascalbaljet** [comment]
I'm open to suggestions, and some improvements were already made in #2146, but I wonder if we can prevent the re-rendering by changing `useForm.ts`.

Even with a simple example like the one below, the _name_ input will re-render when you type into the _company_ input:

```jsx
import { useState } from 'react'

export default () => {
  const [name, setName] = useState('')
  const [company, setCompany] = useState('')

  function submit(event) {
    event.preventDefault()
    //...
  }

  return (
    <form onSubmit={submit}>
      <h1>Form</h1>

      <div>
        <label htmlFor="name">
          Name:
        </label>
        <input
          type="text"
          name="name"
          value={name}
          onChange={(event) => setName(event.target.value)}
        />
      </div>

      <div>
        <label htmlFor="company">
          Company:
        </label>
        <input
          type="text"
          name="company"
          value={company}
          onChange={(event) => setCompany(event.target.value)}
        />
      </div>

      <div>
        <button type="submit">
          Submit
        </button>
      </div>
    </form>
  )
}
```

Memoizing the input components (`const TextInput = React.memo(...)`) will help here, but as you can see, this is just how React works by default.

In `useForm.ts`, all form data is stored in a single object. Refactoring that to store each field separately (and avoid changing the whole object on each update) would be a major refactor, and probably a breaking change.

So please feel free to add to this, correct me if I’m wrong, or share examples of how this could be improved!

**Comment by pascalbaljet** [comment]
I'm closing this issue due to lack of activity. If you're still encountering this and can provide additional details or steps to reproduce, feel free to open a new issue. We'd be happy to take another look. Thanks!

**Comment by chrisk-7777** [comment]
A simple workaround would be updating state on blur instead. I haven't thoroughly tested this, so I don't know if it impacts other things like reset.

```tsx
<label text="Project Name">
    <input
      onBlur={(value) => {
        setData('name', value);
        // validate('name'); <-- doesn't work
        // setTimeout(() => validate('name'); <-- "works" but ugly
      }}
    />
    {invalid('name') ? <div>{errors.name}</div> : null}
</label>
```

However, you can no longer run validate in the onBlur as it won't have access to the latest state. 

The example in https://github.com/inertiajs/inertia/issues/2084#issuecomment-3036578548 is a good reduced case, and the conclusion is correct and that is how React works - updating `company` _will_ trigger a re-render including `name` as its all part of a single component. 

I think this boils down to controlled components (how Inertia useForm works currently) vs uncontrolled components (that is React Hook Form works - see https://www.react-hook-form.com/faqs/ Performance)

Unless the form is really intense, controlled _should_ be fine. 

If there is additional heavy rendering happening as a result of state change happening, you can consider splitting that into a seperate component or use `useDeferredValue` (https://react.dev/reference/react/useDeferredValue#deferring-re-rendering-for-a-part-of-the-ui)

You can also consider debouncing your input, `onChange` debounces the `setData` call
